### ISSUE TITLE
http filters: make the http filter can be configured repeatly and work correctly

### ISSUE THREAD

**Issue by wbpcode** [envoyproxy/envoy#21261]
*Title*: *http filters: make the http filter can be configured repeatly and work correctly*

*Description*:

Although the http filters can be configured repeatly in the `http_filters`, they cannot work perfectly. Because the route configs in the `typed_per_filter_config` are indexed by the filter name. So for the same type filters in the `http_filters`, they can only share the same route-level config.

After #21030, route configs in the `typed_per_filter_config` can be indexed by other keys, for example, `HttpFilter.name`.

This would be a big change. And it would be a long time to deprecate the old behaviours.

**Comment by wbpcode** [comment]
cc @kyessenov

**Comment by wbpcode** [comment]
Some typical scenes:
- multiple `ext_authz` refered by @kyessenov 
- multiple lua scripts for single route
- route config support for wasm filters

**Comment by wbpcode** [comment]
We can achieve this step by step.
- [ ] Ensure the `HttpFilter.name` can be the unique flag in the http filter list.
- [ ] Make the config name and filter name are available for the filters/StreamFilterCallbacks.
- [ ] New API to get route level filter config.
- [ ] Update all the filter used the new API.
- [ ] Switch to use `HttpFilter.name` by runtime flag.
- [ ] Deprecate old behaviours finally.

**Comment by wbpcode** [comment]
Hi, I am trying to land this. 

**Before the first PR, I want to get some inputs from community. Does this make sense for the community?**

(My ultimate goal is per-route or per-vh http filter chain. After complete this work, we can configure same filter repeatedly with different config. And then by #20867, we can enable an array of filters based on route config. For example, for route A, an ext_authz filter backed by auth server A can be enabled, and for route B, an ext_authz filter backed by auth server B can be enabled.)

**Comment by markdroth** [comment]
This looks like a duplicate of #12274.  Can we close this and track the changes there?

Note that there is already some discussion in the other issue about how to handle backward compatibility.

**Comment by wbpcode** [comment]
cc @markdroth you are right.
